Web Inspector: List of breakpoints in Sources tab disappears when inspector reloads

the-chenergy opened the pull request:
#### c6550e29032bf25d2d2dd78a05e29a498d99fc02
<pre>
Web Inspector: List of breakpoints in Sources tab disappears when inspector reloads
<a href="https://rdar.apple.com/123641994">rdar://123641994</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272098">https://bugs.webkit.org/show_bug.cgi?id=272098</a>

Reviewed by Devin Rousso.

The reported bug can be exposed one of three ways, that the list of
breakpoints shown in the source sidebar disappears:
   (1) When you reload the webpage while the inspector is open,
   (2) When you navigate to a different page and come back while the
       inspector is open, and
   (3) When you close and reopen the inspector.

This fix addresses the three ways. Here&apos;s how:

   - For way (1), this happens because:
      a. The source sidebar has the logic to prune &quot;stale&quot; tree
         elements, which are elements that point to resources that no
         longer exists in the current frame. Reloading the webpage
         causes the main resource to change, which triggers this pruning
         logic. (Source: <a href="https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L507-L508)">https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L507-L508)</a>
      b. When adding a breakpoint to show, the source sidebar first
         checks if there already exists a tree element representing
         the breakpoint. So, when the page reloads and its breakpoints
         are added (again), they don&apos;t actually get new corresponding
         tree elements created (Source: <a href="https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1407-L1408).">https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1407-L1408).</a>
      c. The above two operations are in a race. If (a) occurs after
         (b), the _parents_ of the existing breakpoints will get pruned,
         since they now point to removed resources from the old webpage
         and are considered stale. Now, the reason (a) may happen after
         (b) is because NavigationSidebarPanel uses setTimeout to
         coalesce potentially-multiple calls of
         pruneStaleResourceTreeElements into just one, and setTimeout
         may put the call behind schedule. (Source: <a href="https://github.com/WebKit/WebKit/blob/27862391346cb4703757a860478d07f255cf5b57/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L645-L649)">https://github.com/WebKit/WebKit/blob/27862391346cb4703757a860478d07f255cf5b57/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L645-L649)</a>
     So the fix is to remove the coalescence/buffer for calling
     pruneStaleResourceTreeElements and always allow it to be called.
     This does mean we&apos;re potentially pruning many times in a short
     time, but since pruning is only done for the top-level elements in
     the tree outlines, there shouldn&apos;t be a significant number of them
     in most cases. (Pruning nested tree elements when resources are
     sorted By Path is done by the code here: <a href="https://github.com/WebKit/WebKit/blob/bd841a09536799b70842b7e0a7e84890e9853fce/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1073-L1077)">https://github.com/WebKit/WebKit/blob/bd841a09536799b70842b7e0a7e84890e9853fce/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1073-L1077)</a>

   - For way (2), this happens because, when updating the source code
     resource for a breakpoint, which can happen as a result of loading
     a page a second time, that update is skipped unless either the
     old resource or the new resource is null. (Source: <a href="https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L1668-L1669)">https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L1668-L1669)</a>
     The fix is to make it always allow breakpoints&apos; resources to be
     updated, since when reloading the same page the resources will have
     new object references in the frame, and those have no reason to be
     blocked from being updated.

   - For way (3), this happens because:
      a. When the inspector is opened, the DebuggerManager loads the
         list of saved breakpoints from object stores. The manager tries
         to show them in the sidebar by dispatching the BreakpointAdded
         events, but they fail to get added because the breakpoints
         don&apos;t have their linked source code resources yet. (Source: <a href="https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L128-L135)">https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L128-L135)</a>
      b. When a resource gets loaded, it links all loaded breakpoints
         that belong to this resource and show them in the sidebar.
         The breakpoints now succeed to show up because they first get
         linked to their resource. (Source: <a href="https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1187-L1188)">https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1187-L1188)</a>
      c. The above two operations are in a race. If (a) occurs after
         (b), then during (b) there are no breakpoints loaded to link,
         and during (a) the loaded breakpoints can&apos;t get shown without
         linked resources.
     So the fix is, during (b), we also record the resource by its
     `contentIdentifier`, so in case (a) happens afterwards, we can
     look up the loaded resource to link to the loaded breakpoints so
     they can get added.

* Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
(WI.NavigationSidebarPanel):
(WI.NavigationSidebarPanel.prototype.closed):
(WI.NavigationSidebarPanel.prototype.pruneStaleResourceTreeElements):
(WI.NavigationSidebarPanel.prototype._checkOutlinesForPendingViewStateCookie):
(WI.NavigationSidebarPanel.prototype._checkForStaleResourcesIfNeeded): Deleted.
(WI.NavigationSidebarPanel.prototype._checkForStaleResources): Deleted.
   - Remove the logic for coalescing multiple calls into
     pruneStaleResourceTreeElements. Pruning needs to happen immediately
     when conditions are met to prevent stale tree elements from being
     discovered by incoming breakpoints which will be removed by a later
     call to prune.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype._associateBreakpointsWithSourceCode):
* Source/WebInspectorUI/UserInterface/Models/SourceCodeLocation.js:
(WI.SourceCodeLocation.prototype.setSourceCode):
   - Always allow breakpoints&apos; linked-to resource to be updated since
     it might just be a newer reference to the same resource.
   - Move the `console.assert` into `setSourceCode` instead since it
     makes more sense being there instead.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.breakpointsForSourceCode):
   - If existing breakpoints haven&apos;t been loaded yet, record the
     resource by its `contentIdentifier` for later linking the loaded
     breakpoints.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.constructor):
   - When a breakpoint is loaded, see if its resource has already been
     loaded to be linked.

Canonical link: <a href="https://commits.webkit.org/279884@main">https://commits.webkit.org/279884@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93cce2f8f1482f901d9e58e62a9f2797707d9a28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/53746 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/33114 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/6263 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57026 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/4471 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/40606 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/4316 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/57026 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/4471 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/55845 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/40606 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/6263 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57026 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/40606 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/6263 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/2626 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/40606 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/6263 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/58621 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/28913 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/4316 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/58621 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30108 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/6263 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/58621 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12046 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31043 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/29887 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
```diff
@@ -131,6 +132,11 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
                 const key = null;
                 WI.objectStores.breakpoints.associateObject(breakpoint, key, serializedBreakpoint);
 
+                // HACK: Each breakpoint needs to have a linked sourceCode for it to be able to show up in the source sidebar, and the `breakpointsForSourceCode` method normally does the linking job. However, when the inspector opens and the list of breakpoints is first loaded, there is a race between that method runs and this code runs. If that function runs first, breakpoint loaded here won't g ...
```
We don't use `HACK` for comments, and if my question below leads to this being the only place we run this code then we should remove the comment entirely and say the relevant info in the commit message instead.

For comments please also manually line break at some reasonable character limit like 120 chars.

Also in general we should do this work once. Does it work to only do it here instead of `breakpointsForSourceCode`?

the-chenergy replied:
Got it. I'll restyle the comment or remove it, as well as the other long comment I wrote.

> Also in general we should do this work once. Does it work to only do it here instead of breakpointsForSourceCode?

We should absolutely avoid redundant work. Though the call to `_associateBreakpointsWithSourceCode` is actually still only done once for each breakpoint.

This code here and [`breakpointsForSourceCode`](https://github.com/WebKit/WebKit/blob/566221b103c27bb7afc36cd24b303dd5c6535b64/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L413) can be executed in either order when the inspector opens:
   - If this code here happens first, the breakpoints source hasn't been loaded yet and the associating call won't run, which is quite straightforward.
   - If `breakpointsForSourceCode` happens first, the [`_breakpointContentIdentifierMap`](https://github.com/WebKit/WebKit/blob/566221b103c27bb7afc36cd24b303dd5c6535b64/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L422) won't have contained the breakpoints that haven't been added using [`this.addBreakpoint`](https://github.com/WebKit/WebKit/blob/566221b103c27bb7afc36cd24b303dd5c6535b64/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L739-L749) yet which is invoked right here.
      - This is the exact cause of the bug, where the associating call is sometimes done less than once rather than (hopefully) exactly once for each breakpoint.

Let me know if this makes sense or how I could explain this better in the commit message.

dcrousso replied:
honestly the whole idea of having to even do this manual association seems hacky (or at least not ideal).  i would've hoped/preferred that we just do everything based on the `contentIdentifier` instead

with that being said, i think what @patrickangle was maybe trying to get at (and please correct me if i'm wrong) was that this only runs once when Web Inspector is first opened, so it's not ideal to keep `_breakpointSourceCodeContentIdentifierMap` around beyond that point as it'll never be used again

perhaps we should have a `_breakpointsInitialized` flag that's set above which is used in `breakpointsForSourceCode` to prevent anything from being added to  `_breakpointSourceCodeContentIdentifierMap` (which can also be cleared right after this instead of in `_mainResourceDidChange`)?

the-chenergy replied:
Totally, this hack of temporarily remembering the loaded source codes just to link the loaded existing breakpoints is not optimal and makes the code harder to read. I suppose the alternative could be to block the `SourceNavigationSideBar` from being created until the initial list of breakpoints get loaded, which shouldn't take any noticeable time at all but just there to ensure the breakpoint initialization always happens before `breakpointsForSourceCode` is first called. I can do some more digging to see how that can possibly be done or if it'd be even messier trying to do that. What are your thoughts on that?

I did like you and Patrick's suggestion of clearing the hacky source code map as soon as it's no longer useful.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
```diff
@@ -82,6 +82,7 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
         this._breakpointContentIdentifierMap = new Multimap;
         this._breakpointScriptIdentifierMap = new Multimap;
         this._breakpointIdMap = new Map;
+        this._breakpointSourceCodeContentIdentifierMap = new Map;
```
NIT: while this is used for breakpoints, it technically doesn't contain any breakpoints, so id just rename it to `_sourceCodeContentIdentifierMap`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
```diff
@@ -1664,12 +1674,8 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
-            if (!breakpoint.sourceCodeLocation.sourceCode)
-                breakpoint.sourceCodeLocation.sourceCode = sourceCode;
-            // SourceCodes can be unequal if the SourceCodeLocation is associated with a Script and we are looking at the Resource.
-            console.assert(breakpoint.sourceCodeLocation.sourceCode === sourceCode || breakpoint.sourceCodeLocation.sourceCode.contentIdentifier === sourceCode.contentIdentifier);
```
is this `console.assert()` not still useful?  if not, why?

the-chenergy replied:
Ah, this `console.assert()` is still useful because it's still supposed to be true that the `contentIdentifier` of the old and new breakpoints should match. I'll add that part back!

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/SourceCodeLocation.js:
```diff
@@ -276,8 +276,6 @@ WI.SourceCodeLocation = class SourceCodeLocation extends WI.Object
 
     setSourceCode(sourceCode)
     {
-        console.assert((this._sourceCode === null && sourceCode instanceof WI.SourceCode) || (this._sourceCode instanceof WI.SourceCode && sourceCode === null));
```
is this `console.assert()` not still useful?  if not, why?

the-chenergy replied:
This assertion no longer holds because in [`DebuggerManager._associateBreakpointsWithSourceCdoe`](https://github.com/WebKit/WebKit/blob/566221b103c27bb7afc36cd24b303dd5c6535b64/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L1677-L1678), my fix now allows the sources to be updated (not just added or removed) because, when webpage reloads, the sources basically get a new object reference instead of being added for the first time.

Actually, we could instead do the other `console.assert` that you commented on here instead (like we assert if the old and new source code's `contentIdentifier`s match here instead of in that for-loop) if you think that's better.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
```diff
@@ -1404,7 +1404,14 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
     {
-        if (this._breakpointsTreeOutline.findTreeElement(breakpoint))
+        let existingTreeElement = this._breakpointsTreeOutline.findTreeElement(breakpoint);
+        let newSourceCode = breakpoint.sourceCodeLocation.displaySourceCode;
```
not all `WI.Breakpoint` have a `get sourceCodeLocation`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
```diff
@@ -1404,7 +1404,14 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
+        let newSourceCode = breakpoint.sourceCodeLocation.displaySourceCode;
+        let shouldSkipAdding = !!existingTreeElement;
+        if (newSourceCode && breakpoint instanceof WI.JavaScriptBreakpoint && existingTreeElement?.parent.resource !== newSourceCode) {
+            // Normally, we can skip adding a new tree element if there already exists one representing the same breakpoint. But not if that tree element's parent has a different (older) version of source code, which can happen due to a page-reload with the inspector open. In that case, that parent element will be removed by the next `NavigationSidebarPanel.pruneStaleResourceTreeElements()`, so w ...
```
this seems more like a bug with `pruneStaleResourceTreeElements` in that it's not happening soon enough.  i wouldn't be surprised if there's other bugs like this elsewhere that would also benefit from a fix

maybe we need to trigger that in `WI._frameWasAdded` and/or `WI._mainFrameDidChange` instead of or in addition to in `WI._mainResourceDidChange`?

the-chenergy replied:
OK, that does sound probable and reasonable. Let me do some more digging to see what's actually triggering `pruneStaleResourceTreeElements` and why that can happen so late.

the-chenergy replied:
Ah. `NavigationSideBarPanel` [uses `setTimeout`](https://github.com/WebKit/WebKit/blob/27862391346cb4703757a860478d07f255cf5b57/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L645-L649) on `pruneStaleResourceTreeElements` to coalesce multiple incoming calls into one. (Multiple calls might happen due to `_checkForStaleResources` being registered as more than one event listeners.) `setTimeout` might make the pruning call happen later than it should.

I fixed it by allowing `pruneStaleResourceTreeElements` to be called right away and use `setTimeout` to ignore additional calls quickly followed.

dcrousso reviewed (changes requested):
_awesome_ investigations and fixes! 🤩

however, i do think there might be some better ways to approach them 😅

the-chenergy commented on Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
```diff
@@ -485,9 +485,11 @@ WI.NavigationSidebarPanel = class NavigationSidebarPanel extends WI.SidebarPanel
+            this._checkForStaleResourcesTimeoutIdentifier = setTimeout(() => {
+                this._checkForStaleResourcesTimeoutIdentifier = undefined;
+            });
         }
```
Would this check and creation of the timeout identifier be better done in `_checkForStaleResourcesIfNeeded` instead?

dcrousso replied:
this is fine, but i kinda feel like we should modify `Throttler` to support this 🤔

it'd probably involve adding a `fireOnceForThisTask` or something like that (similar to `Debouncer.prototype.delayForTime`)

feel free to do this as a followup if you want

the-chenergy replied:
What would `fireOnceForThisTask` actually do? Like what would be the time it actually delays for?

dcrousso replied:
id imagine it'd do basically what you have this doing right now

the only reason we cant use `Throttler` today is that it's purely time based, but if we assume that every task is at most 16ms (i.e. 60fps) we probably do this and it'd work fine
```js
this._pruneStaleResourceTreeElementsThrottler = new Throttler(() => {
    this.pruneStaleResourceTreeElements();
}, 50);
```

anyways this was just me musing about how to bring this functionality to a more generalized utility since i can imagine it being useful for other things :)

i.e. what you have is totally fine

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
```diff
@@ -131,9 +133,15 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
             }
             this._restoringBreakpoints = false;
+            this._hasLoadedExistingJavaScriptBreakpoints = true;
+            this._sourceCodeContentIdentifierMap.clear();
```
NIT: i wonder if we should `= null` instead since we're never going to use it again

we could even use that instead of `_hasLoadedExistingJavaScriptBreakpoints` in `breakpointsForSourceCode`
```js
this._sourceCodeContentIdentifierMap?.set(sourceCode.contentIdentifier, sourceCode);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/SourceCodeLocation.js:
```diff
@@ -276,11 +276,15 @@ WI.SourceCodeLocation = class SourceCodeLocation extends WI.Object
+            (this._sourceCode === null && sourceCode instanceof WI.SourceCode) ||
+            (this._sourceCode instanceof WI.SourceCode && sourceCode === null) ||
+            (this._sourceCode?.contentIdentifier === sourceCode?.contentIdentifier)
+        );
```
i dont think `Source/WebInspectorUI/Scripts/remove-console-asserts.pl` supports stripping multi-line `console.assert()`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
```diff
@@ -633,20 +635,8 @@ WI.NavigationSidebarPanel = class NavigationSidebarPanel extends WI.SidebarPanel
-
-        // Check on a delay to coalesce multiple calls to _checkForStaleResources.
-        this._checkForStaleResourcesTimeoutIdentifier = setTimeout(this.pruneStaleResourceTreeElements.bind(this));
+        if (this._shouldAutoPruneStaleTopLevelResourceTreeElements && !this._checkForStaleResourcesTimeoutIdentifier)
```
NIT: i think `setTimeout` can return `0` so it's safest to check for `=== undefined`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
```diff
@@ -633,20 +635,8 @@ WI.NavigationSidebarPanel = class NavigationSidebarPanel extends WI.SidebarPanel
-
-    _checkForStaleResources(event)
-    {
-        console.assert(this._shouldAutoPruneStaleTopLevelResourceTreeElements);
```
NIT: i think this can stay as-is since we should only reach here from event listeners added inside `if (this._shouldAutoPruneStaleTopLevelResourceTreeElements)`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
```diff
@@ -1664,12 +1672,8 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
-            console.assert(breakpoint.sourceCodeLocation.sourceCode === sourceCode || breakpoint.sourceCodeLocation.sourceCode.contentIdentifier === sourceCode.contentIdentifier);
-        }
+        for (let breakpoint of breakpoints)
+            breakpoint.sourceCodeLocation.sourceCode = sourceCode;
```
i thought about this a bit more and realized this could be problematic if the same script is loaded more than once (e.g. once in the main page and once in an `<iframe>`) since the `contentIdentifier` is really just an alias for the URL/`sourceURL`

i can see why we do need to do something like this, but it probably should somehow be gated to "once per top-level navigation" or something like that

i admit this is probably a pretty rare occurrence, so it's maybe fine to do as a followup?  at the very least i think we should verify whether my concern is actually valid and if so write a test with a FIXME comment pointing to an open bug

the-chenergy replied:
> i thought about this a bit more and realized this could be problematic if the same script is loaded more than once (e.g. once in the main page and once in an <iframe>) since the contentIdentifier is really just an alias for the URL/sourceURL

What would happen to the source navigation sidebar anyway if there are scripts loaded more than once? Are the newly loaded copies prevented from showing up or do they show up to replace the existing copies?

the-chenergy replied:
> > i thought about this a bit more and realized this could be problematic if the same script is loaded more than once (e.g. once in the main page and once in an <iframe>) since the contentIdentifier is really just an alias for the URL/sourceURL
> 
> What would happen to the source navigation sidebar anyway if there are scripts loaded more than once? Are the newly loaded copies prevented from showing up or do they show up to replace the existing copies?

I see that the source navigation sidebar keeps both copies—they're shown up as two items with the same name. I don't see this being an issue for us when setting the URL though? I feel like setting a breakpoint's `sourceURL` to either of the two copies should both be fine, since the breakpoint should be triggered when the execution hits the line in either file anyway. (I made this test case: [demo.zip](https://github.com/WebKit/WebKit/files/15313253/demo.zip))

While trying that I think I found some issues potentially caused by my fix revolving `pruneStaleResources`. On it now 🏃‍♂️

dcrousso replied:
yeah we'd see the same resource (at least something with the same URL) appear twice

i think the only way this would really be concerning is if somehow the second load returned different content than the first (which is totally possible, but probably not likely)

currently we map breakpoints to the first resource loaded with the given URL, but i think with this change we might end up mapping to the most recently loaded resource instead

again this is probably pretty uncommon so im not too concerned about the difference in behavior (and even then the contents of both resources are likely the same, so we'd just be showing different hierarchical path component navigation items)

maybe we could do some really basic check in the meantime that only does this if we're not in the current session?  something like
```js
let resource = sourceCode;
if (resource instanceof WI.Script)
    resource = resource.resource;
console.assert(!resource || resource instanceof WI.Resource, resource);
for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
    if (parentFrame === WI.networkManager.mainFrame)
        return;
}
```

the-chenergy replied:
> currently we map breakpoints to the first resource loaded with the given URL, but i think with this change we might end up mapping to the most recently loaded resource instead
> 
> ...
> 
> maybe we could do some really basic check in the meantime that only does this if we're not in the current session? something like
> 
> ```js
> let resource = sourceCode;
> if (resource instanceof WI.Script)
>     resource = resource.resource;
> console.assert(!resource || resource instanceof WI.Resource, resource);
> for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
>     if (parentFrame === WI.networkManager.mainFrame)
>         return;
> }
> ```

Wait, I thought the point here is we want to ensure that we still only set the source code if the source code came from the main frame / first loaded frame. The snippet however looks like it's skipping setting the source code if it indeed came from the main frame, which is the opposite of what we want? What am I missing 😅

dcrousso replied:
oh oops yes i wasn't entirely clear.  my apologies 😅

what i meant was that we should only set `breakpoint.sourceCodeLocation.sourceCode = sourceCode;` if the current value for `breakpoint.sourceCodeLocation.sourceCode` is either `null` or not for the main frame.  this way, we'd preserve the existing behavior of only setting the `sourceCode` once (for a given main frame) while still allowing for the new behavior of it changing as the user navigates/refreshes (i.e. across main frames)

so putting it all together id imagine adding this
```js
_isActiveSourceCode(sourceCode)
{
    let resource = sourceCode;

    if (sourceCode instanceof WI.Script) {
        resource = resource.resource;
        if (!resource)
            return this.knownNonResourceScripts.includes(sourceCode);
    }

    console.assert(!resource || resource instanceof WI.Resource, resource);

    for (let parentFrame = resource?.parentFrame; parentFrame; parentFrame = parentFrame.parentFrame) {
        if (parentFrame === WI.networkManager.mainFrame)
            return true;
    }
    return false;
}

_associateBreakpointsWithSourceCode(breakpoints, sourceCode)
{
    if (!this._isActiveSourceCode(sourceCode))
        return;

    this._ignoreBreakpointDisplayLocationDidChangeEvent = true;

    for (let breakpoint of breakpoints) {
        if (this._isActiveSourceCode(breakpoint.sourceCodeLocation.sourceCode))
            continue;

        breakpoint.sourceCodeLocation.sourceCode = sourceCode;
    }

    this._ignoreBreakpointDisplayLocationDidChangeEvent = false;
}
```

im pretty sure this should work even with something like a `JSContext`/`ServiceWorker`/etc. where there are only `WI.Script` without any associated `WI.Resource`, but id suggest double checking just to be sure :)

this is pretty complex, however, so like i mentioned before i think it's fine to have the basic fix for the (likely) common case be done in this PR and the above fix for the (likely) uncommon case be done in a followup

the-chenergy replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=274914 for this potential follow-up fix.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
```diff
@@ -671,7 +648,7 @@ WI.NavigationSidebarPanel = class NavigationSidebarPanel extends WI.SidebarPanel
             return;
 
-        this._checkForStaleResourcesIfNeeded();
+        this.pruneStaleResourceTreeElements();
```
in order to match existing behavior we should probably only do this `if (this._shouldAutoPruneStaleTopLevelResourceTreeElements)`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
```diff
@@ -631,24 +626,6 @@ WI.NavigationSidebarPanel = class NavigationSidebarPanel extends WI.SidebarPanel
-
     _isTreeElementWithoutRepresentedObject(treeElement)
     {
         return treeElement instanceof WI.FolderTreeElement
```
i just noticed this function, and i think a lot of the things in this are no longer valid (e.g. `WI.EventBreakpointTreeElement` does have a `representedObject`), so id suggest looking into this a bit more as an unrelated followup

the-chenergy replied:
Filed at https://bugs.webkit.org/show_bug.cgi?id=274915

the-chenergy committed:
Web Inspector: List of breakpoints in Sources tab disappears when inspector reloads
rdar://123641994
https://bugs.webkit.org/show_bug.cgi?id=272098

Reviewed by Devin Rousso.

The reported bug can be exposed one of three ways, that the list of
breakpoints shown in the source sidebar disappears:
   (1) When you reload the webpage while the inspector is open,
   (2) When you navigate to a different page and come back while the
       inspector is open, and
   (3) When you close and reopen the inspector.

This fix addresses the three ways. Here's how:

   - For way (1), this happens because:
      a. The source sidebar has the logic to prune "stale" tree
         elements, which are elements that point to resources that no
         longer exists in the current frame. Reloading the webpage
         causes the main resource to change, which triggers this pruning
         logic. (Source: https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L507-L508)
      b. When adding a breakpoint to show, the source sidebar first
         checks if there already exists a tree element representing
         the breakpoint. So, when the page reloads and its breakpoints
         are added (again), they don't actually get new corresponding
         tree elements created (Source: https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1407-L1408).
      c. The above two operations are in a race. If (a) occurs after
         (b), the _parents_ of the existing breakpoints will get pruned,
         since they now point to removed resources from the old webpage
         and are considered stale. Now, the reason (a) may happen after
         (b) is because NavigationSidebarPanel uses setTimeout to
         coalesce potentially-multiple calls of
         pruneStaleResourceTreeElements into just one, and setTimeout
         may put the call behind schedule. (Source: https://github.com/WebKit/WebKit/blob/27862391346cb4703757a860478d07f255cf5b57/Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js#L645-L649)
     So the fix is to remove the coalescence/buffer for calling
     pruneStaleResourceTreeElements and always allow it to be called.
     This does mean we're potentially pruning many times in a short
     time, but since pruning is only done for the top-level elements in
     the tree outlines, there shouldn't be a significant number of them
     in most cases. (Pruning nested tree elements when resources are
     sorted By Path is done by the code here: https://github.com/WebKit/WebKit/blob/bd841a09536799b70842b7e0a7e84890e9853fce/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1073-L1077)

   - For way (2), this happens because, when updating the source code
     resource for a breakpoint, which can happen as a result of loading
     a page a second time, that update is skipped unless either the
     old resource or the new resource is null. (Source: https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L1668-L1669)
     The fix is to make it always allow breakpoints' resources to be
     updated, since when reloading the same page the resources will have
     new object references in the frame, and those have no reason to be
     blocked from being updated.

   - For way (3), this happens because:
      a. When the inspector is opened, the DebuggerManager loads the
         list of saved breakpoints from object stores. The manager tries
         to show them in the sidebar by dispatching the BreakpointAdded
         events, but they fail to get added because the breakpoints
         don't have their linked source code resources yet. (Source: https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js#L128-L135)
      b. When a resource gets loaded, it links all loaded breakpoints
         that belong to this resource and show them in the sidebar.
         The breakpoints now succeed to show up because they first get
         linked to their resource. (Source: https://github.com/WebKit/WebKit/blob/c927c492d5b93193bc5cae300f1ca39eb36cd74c/Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js#L1187-L1188)
      c. The above two operations are in a race. If (a) occurs after
         (b), then during (b) there are no breakpoints loaded to link,
         and during (a) the loaded breakpoints can't get shown without
         linked resources.
     So the fix is, during (b), we also record the resource by its
     `contentIdentifier`, so in case (a) happens afterwards, we can
     look up the loaded resource to link to the loaded breakpoints so
     they can get added.

* Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
(WI.NavigationSidebarPanel):
(WI.NavigationSidebarPanel.prototype.closed):
(WI.NavigationSidebarPanel.prototype.pruneStaleResourceTreeElements):
(WI.NavigationSidebarPanel.prototype._checkOutlinesForPendingViewStateCookie):
(WI.NavigationSidebarPanel.prototype._checkForStaleResourcesIfNeeded): Deleted.
(WI.NavigationSidebarPanel.prototype._checkForStaleResources): Deleted.
   - Remove the logic for coalescing multiple calls into
     pruneStaleResourceTreeElements. Pruning needs to happen immediately
     when conditions are met to prevent stale tree elements from being
     discovered by incoming breakpoints which will be removed by a later
     call to prune.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype._associateBreakpointsWithSourceCode):
* Source/WebInspectorUI/UserInterface/Models/SourceCodeLocation.js:
(WI.SourceCodeLocation.prototype.setSourceCode):
   - Always allow breakpoints' linked-to resource to be updated since
     it might just be a newer reference to the same resource.
   - Move the `console.assert` into `setSourceCode` instead since it
     makes more sense being there instead.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.breakpointsForSourceCode):
   - If existing breakpoints haven't been loaded yet, record the
     resource by its `contentIdentifier` for later linking the loaded
     breakpoints.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.constructor):
   - When a breakpoint is loaded, see if its resource has already been
     loaded to be linked.

Canonical link: https://commits.webkit.org/279884@main
